### ISSUE TITLE
Live variables

### ISSUE THREAD

**Issue by dumblob** [til-lang/til#8]
When trying to come up with a concise and explicit way of splitting/copying & merging of streams I came to conclusion that either there'll need to be support for "live variables" as e.g. [Mech lang](https://github.com/mech-lang/mech ) does (where every variable is basically by default live - very cool concept allowing to cut down SLOC count by **about an order of magnitude** in fullstack apps - but the problem currently is performance as this requires quite novel methods of optimization of the generated assembly which is not yet researched enough). 

Or there'll need to be some `select`/`poll`/`kpoll` support and subsequently support for recognition from which stream the value came. This recognition of origin is sometimes done with the typing system, but that might be too cumbersome for Til. In Til I'd probably try to abuse the extraction syntax :wink:.

Maybe even rudimentary things like this could be the starting point (because they can be presumably wrapped by a procedure):

```til
select (range 0 10) (range 5 100) (io.in) | transform x {
  ....
} | case (>x 10) {
  ...
}
```

The latter solution sounds more reasonable for Til. It partially overlaps with [own pipe primitives](https://github.com/til-lang/til/issues/6 ).

**Comment by cleberzavadniak** [comment]
Maybe this could be solved wih processes. I'm starting to feel streams and processes should hang out more together... (Maybe with some syntatic sugar to wrap a command in a new process kind of seamlessly.)

**Comment by dumblob** [comment]
If the plumbing of tubes, ehm streams, is going to be explicit and support both compile-time plumbing (i.e. streams known in compile-time) and runtime plumbing (i.e. streams created at wish in any amount in runtime), then I'm fine with any solution - processes sound generally good. But as always - the devil is in the detail. So if you have any specific API in mind, I'm all ears :wink:.

**Comment by cleberzavadniak** [comment]
With Pids being able to read from and write to streams, that would be easy to implement this:
```tcl
proc f (x) {
  receive | foreach msg {
    send $parent_process "$x / $msg"
  }
}
range 5 | [spawn f whatever] | foreach msg { ... }
```

Someone may think about some kind of syntatic sugar around the `spawn` call, but I would be okay with this explicit version. And it's nice to simply complement Pids instead of creating new syntax.

**Comment by cleberzavadniak** [comment]
Of course, how to reference the parent process is one concern and I personally would prefer to simply call `send` without any Pid instead of injecting a variable into any new scopes.

The last message solution also kind of address the in-Til generators, with:

```tcl
proc g (x) {
  range $x | send
}
[spawn g 7] | foreach msg { ... }
```

**Comment by cleberzavadniak** [comment]
(Except all that would make the Pid be treated as a command. Not sure about this part...)

**Comment by cleberzavadniak** [comment]
About the `select` suggestion, using Pids would feel more natural and I would prefer if the command was very explicit about how it's actually handling data from these multiple streams: maybe you want to "concatenate" data, but maybe you want to "zip" values together or simply "fan out".
```tcl
stream.cat [spawn reader1 "file1.txt"] [spawn reader1 "file2.txt"] | foreach line {...}
stream.zip [spawn reader1 "file1.txt"] [spawn reader1 "file2.txt"] | foreach line {...}
stream.fan_out [spawn reader1 "file1.txt"] [spawn reader1 "file2.txt"] | foreach line {...}
```
(It's relevant to note that **in Til, data is always being pulled, not pushed through pipes**, also.)

`cat` would consume in order, `zip` would always wait until there is a matching on the right side (probably, IDK) and `fan_out` would just consume from whoever has available data at the moment (and that sounds a little bit like a "select").

**Comment by cleberzavadniak** [comment]
1. "I want to add to the same muxer (i.e. to the running | pipe) a third reader which reads yet another file. How will I do it?" - you just stop reading and reorganize things. If you plan to do it, probably is already saving the Pids to variables. Again: in Til streams, data is **pulled**: if you stop pulling, the stream simply stop producing, but is still there to keep producing again later.
2. That's why I created `error.handler ` semantics: in a Unix shell there isn't a nice way either to identify individual errors in a pipeline - all you got is `$?`, basically. And that's not without a reason: the pipeline itself is like a program per se. At least "throwing" an Error allows each pipeline component to convey more context and meaning.
3. Well, there are going to exist plenty of other ways of dealing with UI...

**Comment by dumblob** [comment]
> 1. "I want to add to the same muxer (i.e. to the running | pipe) a third reader which reads yet another file. How will I do it?" - you just stop reading and reorganize things. If you plan to do it, probably is already saving the Pids to variables. Again: in Til streams, data is **pulled**: if you stop pulling, the stream simply stop producing, but is still there to keep producing again later.

Sure and sorry for not being clear with my intentions. My point is two fold.

The pipeline itself seems to not accept any "events" from **outside**. So I can't temporarily interrupt it and "reorganize things" (this is the first problem). I can only stop the whole pipeline and thus can't continue where I left thereafter (this is the second problem).

The second problem is the `trap`/`interrupt` problem which you tried to adress with `error.handler` but that seems to address only "events" from **inside** of the pipeline (which I'm not interested in :wink:).

Any insights how to approach it without `eval` magic (shuffling with pids, orchestrating a lot in each of the participant of the pipeline, etc.)?

**Comment by dumblob** [comment]
Btw. I'm playing with the idea of cutting down on "live variables" functionality and going for a simple **built-in** signal-slot mechanism (simple especially syntactically). I don't know yet.
